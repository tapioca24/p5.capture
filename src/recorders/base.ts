import { EventEmitter } from "events";
import { StrictEventEmitter } from "strict-event-emitter-types";

export type RecorderOptions = {
  width?: number;
  height?: number;
};

type RecorderEmitter = StrictEventEmitter<
  EventEmitter,
  {
    start: void;
    stop: void;
    added: void;
    progress: (progress: number) => void;
    finished: (blob: Blob, filename: string) => void;
    segmented: (blob: Blob, filename: string) => void;
    error: (error: Error) => void;
  }
>;

export type CaptureState = "idle" | "capturing" | "encoding";

export abstract class Recorder extends (EventEmitter as unknown as {
  new (): RecorderEmitter;
}) {
  protected state: CaptureState = "idle";
  protected count = 0;
  protected canvas: HTMLCanvasElement;
  private originalCanvas: HTMLCanvasElement | null = null;
  private shouldResize: boolean = false;

  constructor(canvas: HTMLCanvasElement, options: RecorderOptions = {}) {
    super();

    this.shouldResize = !!(options.width || options.height);
    if (!this.shouldResize) {
      this.canvas = canvas;
      return;
    }
    this.originalCanvas = canvas;
    const { width, height } = this.calcResizeSize(canvas, options);
    this.canvas = document.createElement("canvas");
    this.canvas.width = width;
    this.canvas.height = height;
  }

  get capturedCount() {
    return this.count;
  }

  get captureState() {
    return this.state;
  }

  abstract start(): void;
  abstract stop(): void;
  abstract postDraw(): void;

  protected get canStart() {
    return this.captureState === "idle";
  }

  protected get canStop() {
    return this.captureState === "capturing";
  }

  protected checkStartable() {
    if (!this.canStart) {
      throw new Error("capturing is already started");
    }
  }

  protected checkStoppable() {
    if (!this.canStop) {
      throw new Error("capturing is already started");
    }
  }

  protected reset() {
    this.count = 0;
  }

  protected copyCanvas() {
    if (this.state !== "capturing") return;
    if (!this.shouldResize) return;

    if (!this.originalCanvas) {
      throw new Error("original canvas is not found");
    }

    const ctx = this.canvas.getContext("2d");
    if (!ctx) {
      throw new Error("canvas context is not found");
    }
    ctx.drawImage(
      this.originalCanvas,
      0,
      0,
      this.originalCanvas.width,
      this.originalCanvas.height,
      0,
      0,
      this.canvas.width,
      this.canvas.height,
    );
  }

  protected increment() {
    if (this.state !== "capturing") return;

    this.count++;
    this.emit("added");
  }

  private calcResizeSize(
    original: HTMLCanvasElement,
    options: RecorderOptions,
  ) {
    const { width, height } = options;
    const { width: originalWidth, height: originalHeight } = original;

    if (width && height) {
      return { width, height };
    }

    if (width) {
      return {
        width,
        height: Math.floor(originalHeight * (width / originalWidth)),
      };
    }

    if (height) {
      return {
        width: Math.floor(originalWidth * (height / originalHeight)),
        height,
      };
    }

    return {
      width: originalWidth,
      height: originalHeight,
    };
  }
}
