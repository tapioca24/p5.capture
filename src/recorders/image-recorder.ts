import { ZippableFile, zipSync } from "fflate";
import { Recorder, RecorderOptions } from "@/recorders/base";
import { getDirname, getFilename } from "@/utils";

export type ImageFormat = "png" | "jpg" | "webp";

export type ImageRecorderOptions = RecorderOptions & {
  autoSaveDuration?: number | null;
};

const defaultOptions = {};

type Chunk = {
  index: number;
  filename: string;
  uint8array: Uint8Array;
};

export class ImageRecorder extends Recorder {
  protected tasks: Promise<Chunk>[] = [];
  protected mergedOptions: ImageRecorderOptions;

  constructor(
    canvas: HTMLCanvasElement,
    protected format: ImageFormat,
    options: RecorderOptions = {},
  ) {
    super(canvas, options);
    this.mergedOptions = {
      ...defaultOptions,
      ...options,
    };
  }

  start() {
    this.checkStartable();
    this.state = "capturing";
    this.reset();
    this.emit("start");
  }

  stop() {
    this.checkStoppable();
    this.state = "encoding";
    this.emit("stop");
  }

  async postDraw() {
    try {
      switch (this.captureState) {
        case "capturing":
          const { autoSaveDuration } = this.mergedOptions;
          if (
            autoSaveDuration &&
            this.count > 0 &&
            this.count % autoSaveDuration === 0
          ) {
            this.createSegmentedZip();
          }

          this.copyCanvas();
          const task = this.makeChunk(this.count);
          this.tasks.push(task);

          this.increment();
          break;
        case "encoding":
          this.state = "idle";
          const chunks = await Promise.all(this.tasks);
          const { blob, filename } = this.createZip(chunks);
          this.reset();
          this.emit("finished", blob, filename);
          break;
      }
    } catch (error) {
      if (error instanceof Error) {
        this.emit("error", error);
      }
    }
  }

  protected reset() {
    super.reset();
    this.tasks = [];
  }

  protected makeMimeType(format: ImageFormat) {
    switch (format) {
      case "png":
        return "image/png";
      case "jpg":
        return "image/jpeg";
      case "webp":
        return "image/webp";
      default:
        throw new Error("unknown format");
    }
  }

  protected get qualityOption(): number | undefined {
    return undefined;
  }

  protected async makeBlob() {
    return new Promise<Blob>((resolve, reject) => {
      const mimeType = this.makeMimeType(this.format);
      this.canvas.toBlob(
        (blob) => {
          if (!blob) {
            reject(new Error("failed to make blob"));
            return;
          }
          resolve(blob);
        },
        mimeType,
        this.qualityOption,
      );
    });
  }

  protected async makeChunk(index: number) {
    const numbering = `${index}`.padStart(7, "0");
    const filename = `${numbering}.${this.format}`;

    const blob = await this.makeBlob();
    const arrayBuffer = await blob.arrayBuffer();
    const uint8array = new Uint8Array(arrayBuffer);
    return { index, filename, uint8array };
  }

  protected createZip(chunks: Chunk[]) {
    if (!chunks.length) {
      throw new Error("no chunks");
    }

    const images: Record<string, ZippableFile> = {};
    chunks
      .sort((a, b) => a.index - b.index)
      .forEach((chunk) => {
        images[chunk.filename] = [chunk.uint8array, { level: 0 }];
      });

    const now = new Date();
    const dirname = getDirname(now);
    const zip = zipSync({ [dirname]: images });
    const blob = new Blob([zip], { type: "application/zip" });
    const filename = getFilename(now, "zip");
    return { blob, filename };
  }

  protected async createSegmentedZip() {
    if (!this.tasks.length) {
      throw new Error("no tasks");
    }

    const tasks = [...this.tasks];
    this.tasks = [];

    const chunks = await Promise.all(tasks);
    const { blob, filename } = this.createZip(chunks);
    this.emit("segmented", blob, filename);
  }
}
