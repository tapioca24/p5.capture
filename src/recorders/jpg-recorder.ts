import {
  ImageRecorder,
  ImageRecorderOptions,
} from "@/recorders/image-recorder";

export type JpgRecorderOptions = ImageRecorderOptions & {
  quality?: number;
};

const defaultOptions: JpgRecorderOptions = {
  quality: 0.92,
};

export class JpgRecorder extends ImageRecorder {
  protected mergedOptions: JpgRecorderOptions;

  constructor(canvas: HTMLCanvasElement, options: JpgRecorderOptions = {}) {
    super(canvas, "jpg", options);
    this.mergedOptions = {
      ...defaultOptions,
      ...options,
    };
  }

  protected get qualityOption() {
    return this.mergedOptions.quality;
  }
}
