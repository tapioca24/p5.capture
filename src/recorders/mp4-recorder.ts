import { createH264MP4Encoder, H264MP4Encoder } from "h264-mp4-encoder";
import { Recorder, RecorderOptions } from "@/recorders/base";
import { omitUndefinedProperty } from "@/utils";

type Mp4RecorderOriginalOptions = {
  framerate?: number;
  bitrate?: number;
};

export type Mp4RecorderOptions = RecorderOptions & Mp4RecorderOriginalOptions;

type Mp4RecorderDefaultOptions = {};

const defaultOptions: Mp4RecorderDefaultOptions = {};

export class Mp4Recorder extends Recorder {
  protected encoder: H264MP4Encoder | null = null;
  private mergedMp4Options: Mp4RecorderOriginalOptions &
    Mp4RecorderDefaultOptions;

  constructor(canvas: HTMLCanvasElement, options: Mp4RecorderOptions = {}) {
    super(canvas, options);
    this.mergedMp4Options = {
      ...defaultOptions,
      ...omitUndefinedProperty(options),
    };
  }

  async initialize() {
    const encoder = await createH264MP4Encoder();
    encoder.width = this.canvas.width;
    encoder.height = this.canvas.height;
    encoder.outputFilename = "video.mp4";

    const { framerate, bitrate } = this.mergedMp4Options;
    if (framerate != null) {
      encoder.frameRate = framerate;
      encoder.groupOfPictures = Math.floor(framerate / 2);
    }
    if (bitrate != null) {
      encoder.kbps = bitrate;
    }

    encoder.initialize();
    this.encoder = encoder;
  }

  async start() {
    this.checkStartable();
    this.state = "capturing";
    this.reset();
    this.emit("start");
  }

  async stop() {
    this.checkStoppable();
    this.state = "encoding";
    this.emit("stop");
  }

  async postDraw() {
    try {
      switch (this.captureState) {
        case "capturing":
          if (!this.encoder) {
            throw new Error("encoder is not found");
          }
          this.copyCanvas();
          const imageData = this.getImageData();
          this.encoder.addFrameRgba(imageData);
          this.increment();
          break;
        case "encoding":
          this.state = "idle";
          if (!this.encoder) {
            throw new Error("encoder is not found");
          }
          this.encoder.finalize();
          const uint8Array = this.encoder.FS.readFile(
            this.encoder.outputFilename,
          );
          const blob = new Blob([uint8Array], { type: "video/mp4" });
          const filename = `${this.getBaseFilename(new Date())}.mp4`;
          this.emit("finished", blob, filename);
          break;
      }
    } catch (error) {
      if (error instanceof Error) {
        this.emit("error", error);
      }
    }
  }

  protected getImageData() {
    const { width, height } = this.canvas;

    const ctx = this.canvas.getContext("2d");
    if (ctx) {
      const imageData = ctx.getImageData(0, 0, width, height);
      return imageData.data;
    }

    const gl =
      this.canvas.getContext("webgl2") ?? this.canvas.getContext("webgl");
    if (gl) {
      const pixels = new Uint8Array(width * height * 4);
      gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
      return pixels;
    }

    throw new Error("unsupported context");
  }
}
