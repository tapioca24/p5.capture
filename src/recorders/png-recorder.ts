import {
  ImageRecorder,
  ImageRecorderOptions,
} from "@/recorders/image-recorder";

export type PngRecorderOptions = ImageRecorderOptions & {};

const defaultOptions: PngRecorderOptions = {};

export class PngRecorder extends ImageRecorder {
  protected mergedOptions: PngRecorderOptions;

  constructor(canvas: HTMLCanvasElement, options: PngRecorderOptions = {}) {
    super(canvas, "png", options);
    this.mergedOptions = {
      ...defaultOptions,
      ...options,
    };
  }
}
