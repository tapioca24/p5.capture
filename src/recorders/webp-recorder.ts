import {
  ImageRecorder,
  ImageRecorderOptions,
} from "@/recorders/image-recorder";

export type WebpRecorderOptions = ImageRecorderOptions & {
  quality?: number;
};

const defaultOptions: WebpRecorderOptions = {
  quality: 0.8,
};

export class WebpRecorder extends ImageRecorder {
  protected mergedOptions: WebpRecorderOptions;

  constructor(canvas: HTMLCanvasElement, options: WebpRecorderOptions = {}) {
    super(canvas, "webp", options);
    this.mergedOptions = {
      ...defaultOptions,
      ...options,
    };
  }

  protected get qualityOption() {
    return this.mergedOptions.quality;
  }
}
