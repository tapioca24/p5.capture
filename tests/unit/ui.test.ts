import { describe, it, expect, beforeEach, vi } from "vitest";
import prettier from "prettier";
import { CaptureState } from "@/recorders/base";
import * as ui from "@/ui";

describe("#getFrameCountStr", () => {
  it("returns a zero-padded string", () => {
    expect(ui.getFrameCountStr(1234)).toBe("0001234");
  });
});

describe("#getElapsedTimeStr", () => {
  it("returns a string of elapsed time", () => {
    expect(ui.getElapsedTimeStr(150768, 30)).toBe("1:23:45.6");
    expect(ui.getElapsedTimeStr(100512, 20)).toBe("1:23:45.6");
    expect(ui.getElapsedTimeStr(50256, 10)).toBe("1:23:45.6");
  });
});

describe("#getEncodingProgressStr", () => {
  it("returns 'encoding' with no arguments", () => {
    expect(ui.getEncodingProgressStr()).toBe("encoding");
  });

  it("returns a string with percentages, if progress is specified", () => {
    expect(ui.getEncodingProgressStr(0.5)).toBe("encoding 50%");
    expect(ui.getEncodingProgressStr(0.25)).toBe("encoding 25%");
    expect(ui.getEncodingProgressStr(0.75)).toBe("encoding 75%");
  });
});

describe("#createUi", () => {
  it("inserts the UI into the parent element", async () => {
    const parent = document.createElement("div");
    ui.createUi(parent, { format: "webm", framerate: 30 });
    const html = await prettier.format(parent.innerHTML, { parser: "html" });
    expect(html).toMatchSnapshot();
  });

  describe("event handlers", () => {
    it("calls the onClickRecordButton event handler", () => {
      const onClickRecordButton = vi.fn();
      const parent = document.createElement("div");
      ui.createUi(
        parent,
        { format: "webm", framerate: 30 },
        { onClickRecordButton },
      );
      const button = parent.querySelector<HTMLInputElement>(".p5c-btn");
      button?.dispatchEvent(new Event("click"));
      expect(onClickRecordButton).toHaveBeenCalledTimes(1);
    });

    it("calls the onChangeFormat event handler", () => {
      const onChangeFormat = vi.fn();
      const parent = document.createElement("div");
      ui.createUi(
        parent,
        { format: "webm", framerate: 30 },
        { onChangeFormat },
      );
      const formatSelect =
        parent.querySelector<HTMLSelectElement>(".p5c-format");
      formatSelect?.dispatchEvent(new Event("change"));
      expect(onChangeFormat).toHaveBeenCalledTimes(1);
    });

    it("calls the onChangeFramerate event handler", () => {
      const onChangeFramerate = vi.fn();
      const parent = document.createElement("div");
      ui.createUi(
        parent,
        { format: "webm", framerate: 30 },
        { onChangeFramerate },
      );
      const framerateInput =
        parent.querySelector<HTMLInputElement>(".p5c-framerate");
      framerateInput?.dispatchEvent(new Event("change"));
      expect(onChangeFramerate).toHaveBeenCalledTimes(1);
    });
  });

  describe("#updateUi", () => {
    let parent: HTMLDivElement;
    let updateUi: ReturnType<typeof ui.createUi>["updateUi"];

    beforeEach(() => {
      parent = document.createElement("div");
      updateUi = ui.createUi(parent, {
        format: "webm",
        framerate: 30,
      }).updateUi;
    });

    describe("container", () => {
      let container: HTMLDivElement | null;

      beforeEach(() => {
        container = parent.querySelector<HTMLDivElement>(".p5c-container");
      });

      it("has a container with the idle class in the initial state", () => {
        expect(container!.classList.contains("idle")).toBe(true);
        expect(container!.classList.contains("capturing")).toBe(false);
        expect(container!.classList.contains("encoding")).toBe(false);
      });

      it.each<{ state: CaptureState; expected: CaptureState }>([
        { state: "idle", expected: "idle" },
        { state: "capturing", expected: "capturing" },
        { state: "encoding", expected: "encoding" },
      ])(
        "has a container with a $expected class in $state state",
        ({ state, expected }) => {
          const unexpected = ["idle", "capturing", "encoding"].filter(
            (state) => state !== expected,
          );

          updateUi(state, 0, 30);
          expect(container!.classList.contains(expected)).toBe(true);
          unexpected.forEach((state) => {
            expect(container!.classList.contains(state)).toBe(false);
          });
        },
      );
    });

    describe("record button", () => {
      let recordButton: HTMLButtonElement | null;

      beforeEach(() => {
        recordButton = parent.querySelector<HTMLButtonElement>(".p5c-btn");
      });

      it("has a enabled button in the initial state", () => {
        expect(recordButton!.disabled).toBeFalsy();
      });

      it.each<{ state: CaptureState; expected: boolean }>([
        { state: "idle", expected: false },
        { state: "capturing", expected: false },
        { state: "encoding", expected: true },
      ])(
        "hasn a disabled button only in encoding state",
        ({ state, expected }) => {
          updateUi(state, 0, 30);
          expect(recordButton!.disabled).toBe(expected);
        },
      );
    });

    describe("counter", () => {
      let counter: HTMLSpanElement | null;

      beforeEach(() => {
        counter = parent.querySelector<HTMLSpanElement>(".p5c-counter");
      });

      it("has a empty counter in the initial state", () => {
        expect(counter!.textContent).toBe("");
      });

      it.each<{
        state: CaptureState;
        count: number;
        framerate?: number;
        progress?: number;
        expected: string;
      }>([
        { state: "idle", count: 123, expected: "0000123" },
        { state: "idle", count: 123, framerate: 30, expected: "0:00:04.1" },
        { state: "idle", count: 123, progress: 0.5, expected: "0000123" },
        {
          state: "idle",
          count: 123,
          framerate: 30,
          progress: 0.5,
          expected: "0:00:04.1",
        },
        { state: "capturing", count: 123, expected: "0000123" },
        {
          state: "capturing",
          count: 123,
          framerate: 30,
          expected: "0:00:04.1",
        },
        {
          state: "capturing",
          count: 123,
          progress: 0.5,
          expected: "0000123",
        },
        {
          state: "capturing",
          count: 123,
          framerate: 30,
          progress: 0.5,
          expected: "0:00:04.1",
        },
        { state: "encoding", count: 123, expected: "encoding" },
        {
          state: "encoding",
          count: 123,
          framerate: 30,
          expected: "encoding",
        },
        {
          state: "encoding",
          count: 123,
          progress: 0.5,
          expected: "encoding 50%",
        },
        {
          state: "encoding",
          count: 123,
          framerate: 30,
          progress: 0.5,
          expected: "encoding 50%",
        },
      ])(
        "has a count that has a text content corresponding to arguments",
        ({ state, count, framerate, progress, expected }) => {
          updateUi(state, count, framerate, progress);
          expect(counter!.textContent).toBe(expected);
        },
      );
    });

    describe("format select", () => {
      let formatSelect: HTMLSelectElement | null;

      beforeEach(() => {
        formatSelect = parent.querySelector<HTMLSelectElement>(".p5c-format");
      });

      it("has a format select with webm as value in the initial state", () => {
        expect(formatSelect!.value).toBe("webm");
      });

      it.each<{ state: CaptureState; expected: boolean }>([
        { state: "idle", expected: false },
        { state: "capturing", expected: true },
        { state: "encoding", expected: true },
      ])(
        "has a enable format select only in idle state",
        ({ state, expected }) => {
          updateUi(state, 0, 30);
          expect(formatSelect!.disabled).toBe(expected);
        },
      );
    });

    describe("framerate input", () => {
      let framerateInput: HTMLInputElement | null;

      beforeEach(() => {
        framerateInput =
          parent.querySelector<HTMLInputElement>(".p5c-framerate");
      });

      it("has a framerate input with 30 as value in the initial state", () => {
        expect(framerateInput!.valueAsNumber).toBe(30);
      });

      it.each<{ state: CaptureState; expected: boolean }>([
        { state: "idle", expected: false },
        { state: "capturing", expected: true },
        { state: "encoding", expected: true },
      ])(
        "has a enable framerate input only in idle state",
        ({ state, expected }) => {
          updateUi(state, 0, 30);
          expect(framerateInput!.disabled).toBe(expected);
        },
      );
    });
  });
});
