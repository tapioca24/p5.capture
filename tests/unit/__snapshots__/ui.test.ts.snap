// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#createUi > inserts the UI into the parent element 1`] = `
"<div class="p5c-container idle">
  <div class="p5c-options">
    <label for="p5c-format">format</label
    ><select id="p5c-format" class="p5c-format">
      <option value="webm">WebM</option>
      <option value="gif">GIF</option>
      <option value="mp4">MP4</option>
      <option value="png">PNG</option>
      <option value="jpg">JPG</option>
      <option value="webp">WebP</option></select
    ><label for="p5c-framerate">framerate</label
    ><input id="p5c-framerate" class="p5c-framerate" type="number" min="1" />
  </div>
  <div class="p5c-main">
    <button class="p5c-btn"></button><span class="p5c-counter"></span>
  </div>
</div>
"
`;
